### ISSUE TITLE
Testing and continuous integration (CI)

### ISSUE THREAD

**Issue by Chris-Petty** [sussol/mobile#442]
Description:
Currently there are no unit tests built into mobile, and no testing pipeline other than hands on testing by the sussol team. The lack of this is a hinderance to potential open source developers out there looking to chip in, as well as our team itself.
This should also make future refactoring tasks safer.
I'd suggest only taking it as far a >90% coverage with no CI builds. Just PR unit test checks should be sufficient.
Interface snapshot testing would be good, but I think lower priority than core logic unit tests.

Comments:
Testing sync will be a bit of a mission, #407 might lay some ground work for doing that in a effective manner. Sync would benefit largely from testing, as regressions in this core piece of the app are very painful!

It will add development overhead, but only upfront; most of the cost is already there, just hidden in potential regression and repetitive manual testing. It'll save time in that manner in exchange for some additional well spent coding time. Initial cost might seem high, but only because of built up technical debt.

I'd say that with a good testing approach set up, this would be a good piece to give a summer intern a shot at, churning out some of the test code.

**Comment by joshxg** [comment]
@Chris-Petty @andreievg @wlthomson @mark-prins @nmadruga @DhanyaHerath 

I am looking at upgrading jest and writing some unit tests for the core vaccine module functions. I think this will only take slightly more time than doing some white box testing on this core functionality manually (which would be done in the build test).

I am looking to do this quite differently than how I have written tests for the integration service's previously. I've been convinced that development code should be written in a different style than production code:

- As simple as possible and as explicit as possible (No looping over multiple functions to run the same test, just explicitly writing multiple tests, if needed). Duplication, excessive variables and minimal-to-no factoring
- As detailed as possible within an individual test (Such that one could look at only the test and function and have enough information to diagnose the problem).
-  All code in one place: No files of `testData`, or anything like that - should be able to go to the test and see exactly what you need without having to navigate around the file system

Anything anyone has to add?

**Comment by Chris-Petty** [comment]
@joshxg I agree with this change in style, unit tests should be just about as imperative as possible with linear/flat control flows.

Gonna throw in here that Jest and Jasmine basically have the same API, Jasmine has some good docs indicating style/structuring https://jasmine.github.io/tutorials/your_first_suite

FB use Jest for testing React native themselves (large part of why they developed Jest anyway). Here is the guide on snapshot testing for RN UI https://jestjs.io/docs/en/tutorial-react-native
